tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,280 @@
+use sharded_slab::{Guard, Slab};
+
+use crate::{
+    fmt::span::SpanStack,
+    registry::{LookupSpan, SpanData},
+};
+use serde_json::{json, Value};
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    convert::TryInto,
+    fmt,
+    sync::{
+        atomic::{fence, AtomicUsize, Ordering},
+        Arc, Mutex,
+    },
+};
+use tracing_core::{
+    dispatcher,
+    field::{FieldSet, Visit},
+    span::{self, Id},
+    Event, Field, Interest, Metadata, Subscriber,
+};
+
+#[derive(Debug)]
+pub struct Registry {
+    spans: Arc<Slab<Data>>,
+}
+
+thread_local! {
+    static CONTEXT: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+#[derive(Debug)]
+pub struct Data {
+    metadata: &'static Metadata<'static>,
+    parent: Option<Id>,
+    children: Vec<Id>,
+    ref_count: AtomicUsize,
+    // TODO(david): get rid of these
+    values: Mutex<HashMap<&'static str, Value>>,
+}
+
+impl Drop for Data {
+    fn drop(&mut self) {
+        // We have to actually unpack the option inside the `get_default`
+        // closure, since it is a `FnMut`, but testing that there _is_ a value
+        // here lets us avoid the thread-local access if we don't need the
+        // dispatcher at all.
+        if self.parent.is_some() {
+            dispatcher::get_default(|subscriber| {
+                if let Some(parent) = self.parent.take() {
+                    let _ = subscriber.try_close(parent);
+                }
+            })
+        }
+    }
+}
+
+// === impl Registry ===
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self {
+            spans: Arc::new(Slab::new()),
+        }
+    }
+}
+
+struct RegistryVisitor<'a>(&'a mut HashMap<&'static str, Value>);
+
+impl<'a> Visit for RegistryVisitor<'a> {
+    // TODO: special visitors for various formats that honeycomb.io supports
```

Comment:
?

I don't think that's up to the registry...in fact, I think we'll want to remove the visitor that records fields from the registry entirely. A honeycomb subscriber would provide its own visitor and use the extensions interface.

Reply:
oh yeah, this needs to be removed and placed into one of the layers (or, operate on the extensions).